provide support for the ST S2LP 802.15.4 Radio

nikooiko opened the pull request:
Hello everyone, in this PR the target is to provide support for the [ST S2LP 802.15.4 Radio](https://www.st.com/en/wireless-transceivers-mcus-and-modules/s2-lp.html).

It's a work in progress but it's at a working state and can be the base for the enhancements/upgrades discussion.

Included in this merge request:
* Driver for the S2LP 802.15.4 radio
* The expansion of the ST external support library with the [x-cube-subg1](https://www.st.com/content/st_com/en/products/embedded-software/mcu-mpu-embedded-software/stm32-embedded-software/stm32cube-expansion-packages/x-cube-subg1.html) expansion package, written by ST Microelectronics.
* The definition of the [x_nucleo_s2868a1](https://www.st.com/en/ecosystems/x-nucleo-s2868a1.html) expansion module.

Current TODO list:
- [ ] Support MTU > 127 bytes. Currently the driver works only with short MTU=127 bytes, but based on datasheet, the device should support MTU up to 65535 with basic packet format.
- [ ] Correctly configure radio Rx State to optimize power consumption (Currently draws around 8mA but can be optimized with LDC and Rx Sniff to operate with lower consumption).
- [x] Rework Rx Thread main function (s2lp_rx) to simplify and avoid unneeded operations.
- [x] Implement s2lp_set_channel and s2lp_get_channel_count functions.
- [x] Add data rate configuration option (currently operates at 38.4Kbps).
- [x] Add frequency band configuration option  (currently operates at 868Mhz).
- [x] BUG: There is a case where the TX operation locks because IRQ_TX_DATA_SENT is not correctly sampled (missed by the polling operation).

Possible changes:
- [ ] The radio supports HW filter and HW auto-ack mechanisms. Is there any interest in implementing them?
- [ ] X cube subg1 software expects little-endian architectures. Should be patched to also provide support for big-endian architectures?
- [ ] The current driver only supports [S2LP](https://www.st.com/en/wireless-transceivers-mcus-and-modules/s2-lp.html) radio but can be expanded to also support the [ST SPIRIT1 802.15.4 radio](https://www.st.com/en/wireless-transceivers-mcus-and-modules/spirit1.html).
- [ ] The current driver works with a static config for the radio params. Should be upgraded to be configurable through Kconfig workflow? Is there any interest in this?
- [ ] Replace basic packet format with the 802.15.4 packet format?

Signed-off-by: Nikos Oikonomou nikoikonomou92@gmail.com

nikooiko committed:
ieee802154: added support for s2lp radio module

Added driver for the s2lp radio modules

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

erwango commented on ext/hal/st/stm32cube_subg1/README:
```diff
@@ -0,0 +1,36 @@
+   External
+
+License:
+   BSD-3-Clause
```
This would be great, but where did you found the info?
Looking to proposed file, I can't find this clearly stated

nikooiko replied:
This is a placeholder that I forgot to remove, when I copied the format of the other st-cube libraries (e.g. stm32f0xx)

nikooiko replied:
The actual licence for this lib is this: https://my.st.com/content/ccc/resource/legal/legal_agreement/license_agreement/group0/62/6a/48/15/b4/12/47/7a/DM00367782/files/DM00367782.pdf/jcr:content/translations/en.DM00367782.pdf

nikooiko replied:
For reference you can find the licence link here: https://my.st.com/content/my_st_com/en/products/embedded-software/mcu-mpu-embedded-software/stm32-embedded-software/stm32cube-expansion-packages/x-cube-subg1.license=1557757423596.product=X-CUBE-SUBG1.version=3.1.0.html#

erwango replied:
Ok. Well this is actually a License agreement, and not exactly a License.
It says that we should use components of the package according to the License they disclaim.
So either this is II and we're good:
- FOR THE LICENSED SOFTWARE OR ANY PART THEREOF WHICH CONTAINS OPEN SOURCESOFTWARE: THE APPLICABLE OPEN SOURCE TERMS SHALL APPLY.

Either this is III, and we have an issue:
-  THE LICENSED SOFTWARE OR ANY PARTS THEREOF MAY (I) REQUIRE LICENSES FROMTHIRD PARTIES CLAIMING INTELLECTUAL PROPERTY RIGHTS COVERING USE ORIMPLEMENTATION OF THE LICENSED SOFTWARE OR (II) BE BASED ON INDUSTRYRECOGNIZED STANDARDS OR SOFTWARE PROGRAMS PUBLISHED BY INDUSTRYRECOGNIZED STANDARDS BODIES AND CERTAIN THIRD PARTIES MAY CLAIM TO OWNINTELLECTUAL PROPERTY RIGHTS THAT COVER IMPLEMENTATION OR USE OF THOSESTANDARDS. YOU AGREE THAT YOU ARE RESPONSIBLE FOR OBTAINING ANY SUCH LICENSEWHICH MAY BE NEEDED, AND NO SUCH LICENSE IS PROVIDED BY ST OR ITS AFFILIATES TOYOU .SUCH THIRD PARTY INTELLECTUAL PROPERTY RIGHTS ARE NOT SUBJECT TO THETERMS OF THIS AGREEMENT TO THE EXTENT THE TERMS OF THIS AGREEMENT ARE INCONFLICT WITH SUCH APPLICABLE THIRD PARTY INTELLECTUAL PROPERTY RIGHTS.

So we need to find out which one applies here.

nikooiko replied:
Is there something I can do to help with this?

erwango replied:
I don't think so. I'll check internally and will keep you posted.
It might take a while, so review should continue on pure code part zephyr integration during this time.

erwango replied:
I have the confirmation that ST proposes this library in a permissive License (BSD-3 Clause or alike). So there is nothing that prevent its use in Zephyr.
We just need to get an updated version of the lib with updated header files reflecting this.

erwango reviewed (changes requested):
Thanks for this work.
For a start, I will let others commenting on the code part, and I will focus on the license statement.
We need to ensure the license of the files you extracted from the package is compatible with use in Zephyr.
I will nack this change until it is proven we're safe on that side, but it will (should) not prevent others to review the other parts of the change meanwhile.

dbkinder commented on boards/shields/x_nucleo_s2868a1/doc/index.rst:
```diff
@@ -0,0 +1,60 @@
+ - RoHS compliant
+
+
+More information about X-NUCLEO-S2868A1 can be found here:
```
This renders better without the bullet point:
```
More information about X-NUCLEO-S2868A1 can be found in the
`X-NUCLEO-S2868A1 data sheet`_.
```

nikooiko replied:
this is resolved

dbkinder reviewed (changes requested):
Doc is OK so far (with one correction).  Is a usage example coming later?

nikooiko commented:
With the provided overlay files we can execute the net > echo server/client samples. It doesn't need to showcase anything different than any other 802.15.4 device.

Of course, if you believe that a custom sample is required I can either provide a new net-based example (similar to echo), or a custom low-level sample with `CONFIG_IEEE802154_RAW_MODE` enabled. If you prefer the latter, I could provide a stress test sample to evaluate the performance of the radio and the provided driver.

erwango commented on samples/net/sockets/echo_client/overlay-s2868a1.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_GPIO=y
```
Once again, please follow up on https://github.com/zephyrproject-rtos/zephyr/pull/14057/commits
shield overlays would now sit under boards/shields, so they are not tight to a sample

nikooiko commented on boards/shields/x_nucleo_s2868a1/x_nucleo_s2868a1.overlay:
```diff
@@ -0,0 +1,20 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&spi1 {
```
so it will be `arduino_spi: &spi1`?

erwango replied:
Shield overlays should refer to generic nodes label, such as arduino_spi.
These alternate labels (overridding nodes in dts grammar) are defined in boards, cf nucleo_f429zi.dts.
This allows shield to be compatible with various boards that would have a different definition of arduino_spi.

erwango commented:
recheck

dbkinder reviewed (approved):
+1 for docs

erwango commented:
@nikooiko, you'll find in https://www.st.com/en/embedded-software/x-cube-subg1.html a 3.1.1 version of the package that includes BSD-3 zephyr compatible License. Please use this package.
Also, due to move of `ext\` components to zephyr modules, you'll need to push part of your changes in following git repo: https://www.st.com/en/embedded-software/x-cube-subg1.html

Thanks

nikooiko commented:
@erwango thanks for the ping!
I'll change them when I find the time.

nikooiko commented:
> @nikooiko , any plan for this work?

Sorry for the inactivity. I've been too busy with other things and I haven't found the time to work with this driver. Hopefully, a colleague of mine will start working on this radio so he might have the time to continue the driver.

erwango commented:
@nikooiko, thanks for the feedback. We're in the process of cleaning up the PR base. I'm tagging the PR as "stale", it might also be closed later on. But don't worry, it can still be re-opened, and I'll be happy to continue the review.

galak commented:
Closing PR as its been marked stale for several months. Feel free to re-open and update if desired.

nikooiko committed:
west.yml: Update st revision

st: update to latest ST package revision '44d5d70'.

Includes ST S2LP Library.

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko committed:
boards: shields: support for x_nucleo_s2868a1

Added a new shield for the x_nucleo_s2868a1 expansion module

Signed-off-by: Nikos Oikonomou <nikoikonomou92@gmail.com>

nikooiko commented:
Hello everyone, I updated this merge request (with all the new changed APIs, e.g. GPIO, DT) and following the new STLib structure, I also introduced this related pull request https://github.com/zephyrproject-rtos/hal_st/pull/6.

erwango commented:
@nikooiko Good to see your still working on this. 
Before pushing again on this PR, you'll need to re-open it (otherwise you won't be able to reopen it again).

erwango commented:
@nikooiko There were a bunch of TODO and squares to tick in your PR description. Is that still valid? If no, would you mind providing an updated status ?

nikooiko commented:
These are still valid as the core of the driver is still the same. The reason I initially opened this pull request is mainly for starting a discussion for this driver, and possibly get some ideas for it.

In general, it's a simple but working prototype driver with many open issues, and what I'll try to do first is to verify that it's working without problems and then I'll continue with those bullets, and at the same time, I'll try to simplify the flow as much as possible.

The first bullet that I'll try to change is the one for the packet format replacement (Basic vs 802.15.4), as I think that the latter is more appropriate/generic.

erwango commented on west.yml:
```diff
@@ -72,7 +72,7 @@ manifest:
       path: modules/hal/silabs
     - name: hal_st
-      revision: b52fdbf4b62439be9fab9bb4bae9690a42d2fb14
+      revision: 44d5d70b09aeae37c225526e46baa06b5fc79885
```
While it isn't yet merged, you should use the following reference:
```suggestion
      revision: /pull/6/head
```

nikooiko replied:
I updated it, but CI still fails.

erwango replied:
@nikooiko my bad this should be : 
```suggestion
revision: pull/6/head
```

erwango reviewed (changes requested):
@nikooiko First change in order to get CI running correctly

erwango commented:
Definitely there is something wrong in the way we're calling hal_st

erwango commented:
@nikooiko CI is still failed ... but hal update step is passed! Some CI checks to fix and you should be done (with this part ;-))

erwango commented:
@nikooiko Seems you're out of luck with IC, this timeit is not related with your change. You can have a try to rebase and force push to see if this is getting better.
Regarding the driver part, it would be nice if @tbursztyka or @jukkar could have a look.

nikooiko commented:
> @nikooiko Seems you're out of luck with IC, this timeit is not related with your change. You can have a try to rebase and force push to see if this is getting better.
> Regarding the driver part, it would be nice if @tbursztyka or @jukkar could have a look.

Guys here is some feedback about the driver's state/progress:

As I mentioned above I investigated replacing basic packet format with 802.15.4, but in the end, I found no reason to do so.

I tried to use the echo_client/echo_server samples but the nodes were unable to communicate so I'll need to investigate it further. At the same time I started working with a custom sample that uses 802.15.4 raw communication ([s2868a1 sample](https://github.com/nikooiko/zephyr/blob/st-subg1-s2lp-sample/samples/shields/x_nucleo_s2868a1/)) to stress test and showcase the radio operation.  

In parallel my next target is to implement the s2lp_set_channel and s2lp_get_channel_count functions.

As I'm still working on this driver, any feedback would be much appreciated to produce a better result.

nikooiko commented:
Guys, unfortunately, I was again forced to pause the development of this request. Until I have the time to continue with it or someone else steps up I'm closing this pull request.

roshangeoroy commented:
Hi @nikooiko ,
I _recently_ worked with S2LP on a project, so if you're still interested, I’d be happy to help finish this driver. Let me know if we can reopen the pull request before it’s too late and S2LP **fades** from my memory!

roshangeoroy commented:
Hi @nikooiko , 
Just following up on this. 

@erwango, do you think we can move forward with reopening the PR?

nikooiko commented:
Hi @roshangeoroy, unfortunately, I stopped working on this for quite some time now, so I won't be able to assist you with this.
Feel free to take over ;)

roshangeoroy commented:
Alright then! I’ll likely start a new PR and link this one there.

For the initial tasks, I’ll be focusing on:
- Shifting from xcube-subg1 to  [xcube-subg2](https://github.com/STMicroelectronics/x-cube-subg2).
- Adding the new shield for the S2868A2 board (since S2868A1 is no longer recommended for new designs).

Some of the possible TODOs I’ll tackle first include:
- Utilizing the hardware MAC features (packet format, filtering, autoack).
- Optimizing power consumption (LDC and Sniff mode).
